packages/astro/src/cli/add.ts

```diff
@@ -0,0 +1,361 @@
+	const output = await generate(ast, fileURLToPath(configURL));
+
+	if (input === output) {
+		return UpdateResult.none;
+	}
+
+	const message = `\n${boxen(
+		diffLines(input, output)
+			.map((change) => {
+				let lines = change.value.split('\n').slice(0, -1); // remove latest \n
+
+				if (change.added) lines = lines.map((line) => green(`+ ${line}`));
+				else if (change.removed) lines = lines.map((line) => red(`- ${line}`));
+				else lines = lines.map((line) => dim(`  ${line}`));
+
+				return lines.join('\n');
+			})
+			.join('\n'),
+		{ margin: 0.5, padding: 0.5, borderStyle: 'round', title: configURL.pathname.split('/').pop() }
+	)}\n`;
+
+	info(logging, null, `\n  ${magenta('Astro will update your configuration with these changes...')}\n${message}`);
+
+	const response = await prompts({
+		type: 'confirm',
+		name: 'updateConfig',
+		message: 'Continue?',
+		initial: true,
+	});
+
+	if (response.updateConfig) {
+		await fs.writeFile(fileURLToPath(configURL), output, { encoding: 'utf-8' });
+		debug('add', `Updated astro config`);
+		return UpdateResult.updated;
+	} else {
+		return UpdateResult.cancelled;
+	}
+}
+
+async function getInstallIntegrationsCommand({ integrations, cwd = process.cwd() }: { integrations: IntegrationInfo[]; cwd?: string }): Promise<string | null> {
+	const pm = await preferredPM(cwd);
+	debug('add', `package manager: ${JSON.stringify(pm)}`);
+	if (!pm) return null;
+
+	let dependenciesList = integrations
+		.map<[string, string | null][]>((i) => [[i.packageName, null], ...i.dependencies])
+		.flat(1)
+		.filter((dep, i, arr) => arr.findIndex((d) => d[0] === dep[0]) === i)
+		.map(([name, version]) => (version === null ? name : `${name}@${version}`))
+		.sort()
+		.join(' ');
+
+	switch (pm.name) {
+		case 'npm':
+			return 'npm install --save-dev ' + dependenciesList;
+		case 'yarn':
+			return 'yarn add --dev ' + dependenciesList;
+		case 'pnpm':
+			return 'pnpm install --save-dev ' + dependenciesList;
+		default:
+			return null;
+	}
+}
+
+async function tryToInstallIntegrations({
+	integrations,
+	cwd = process.cwd(),
+	logging,
+}: {
+	integrations: IntegrationInfo[];
+	cwd?: string;
+	logging: LogOptions;
+}): Promise<UpdateResult> {
+	const cmd = await getInstallIntegrationsCommand({ integrations, cwd });
+
+	if (cmd === null) {
+		info(logging, null);
+		return UpdateResult.none;
+	} else {
+		const message = `\n${boxen(cyan(cmd), { margin: 0.5, padding: 0.5, borderStyle: 'round' })}\n`;
+		info(logging, null, `\n  ${magenta('Astro will run the following command to install...')}\n${message}`);
+		const response = await prompts({
+			type: 'confirm',
+			name: 'installDependencies',
+			message: 'Continue?',
+			initial: true,
+		});
+
+		if (response.installDependencies) {
+			const spinner = ora('Installing dependencies...').start();
+			try {
+				await new Promise((resolve, reject) => {
+					exec(cmd, (err, stdout, stderr) => {
```

Comment:
I know its another dep for this feature, but I'm a strong fan of `execa` over `exec` for reliability, which I think is at a premium here. In this case I'd rather start with the confidence choice and then optimize later.

For example, I was going to suggest piping the output here as a stream, so that it doesn't buffer and then throw like this. Also, combining stdout and sterr together, since I think logs can happen in either stream at any time, in both success and failure. Also, not sure if `cwd` is handled here.

Not sure how to do this with exec, but with execa its:

```js
import { execaCommand } from 'execa';

await execaCommand(cmd, { cwd, stdout: 'inherit', stderr: 'inherit' });
```